### PR TITLE
Support complex numbers in serial linear solvers, add BLAS zdot

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Adds complex numbers support to the serial linear solvers, which are used as preconditioners in the parallel linear solvers. Most notably, extends the GMRES algorithm to complex arithmetics, which involves the inner product on the vector space of complex numbers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@wthrowe would you be willing to review this one?

File: tests/Unit/Utilities/TypeTraits/Test_GetFundamentalType.cpp [hunk 1]
```diff
@@ -21,3 +21,17 @@ static_assert(
...
               "Failed testing get_fundamental_type");
 // [get_fundamental_type]
+
+static_assert(
+    std::is_same_v<
+        typename tt::get_complex_or_fundamental_type_t<std::array<double, 2>>,
+        double>);
+static_assert(std::is_same_v<typename tt::get_complex_or_fundamental_type_t<
+                                 std::vector<std::complex<int>>>,
+                             std::complex<int>>);
```
Comment (wthrowe) [inline tests/Unit/Utilities/TypeTraits/Test_GetFundamentalType.cpp]: `std::complex` is only guaranteed to work for floating-point types.

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_BuildMatrix.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ SPECTRE_TEST_CASE("Unit.LinearSolver.Serial.BuildMatrix",
   {
     INFO("Build a simple dense matrix");
     const blaze::DynamicMatrix<double> matrix{{4., 1.}, {3., 1.}};
-    const helpers::ApplyMatrix linear_operator{matrix};
+    const helpers::ApplyMatrix<double> linear_operator{matrix};
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/LinearSolver/Test_BuildMatrix.cpp:45]: Wrong commit?

**Review by wthrowe — CHANGES_REQUESTED** [review]
Looks overall OK, but it's not compiling, and it looks like the intermediate commits won't compile even after the overall problems are fixed.

File: tests/Unit/Utilities/TypeTraits/Test_GetFundamentalType.cpp (hunk 1 above)
Comment (nilsvu) [reply tests/Unit/Utilities/TypeTraits/Test_GetFundamentalType.cpp]: Changed to `double`

**Comment by wthrowe** [comment]
Still doesn't compile.

**Comment by nilsvu** [comment]
Should be fixed now!

**Review by wthrowe — APPROVED** [review]

**Pull request merged**